### ISSUE TITLE
Audio instability/disturbances with Jamulus

### ISSUE THREAD

**Issue by gvermag22** [corrados/jamulus#396]
We have been testing with Jamulus a lot for a non profit organization that does a lot of music. The technology is promising, but flaky. We have seen several issues with audio and compatibility with mics. Yesterday, we were testing with one singer/instrument player in Austin, TX on a google fiber network, whereas two of us were in Fremont, CA using Comcast for internet. The server was running in a data center in SFO, CA in my home. The delays for the fremont folks was 60ms and the delay from Austin was 80-85ms (due to the fact that it was in a data center, the delay reduced due to closer access to internet backbone). 

However, there were significant audio issues. The audio was buffering or making sounds like phat-phat. We also heard sparking sounds. We had to re-connect many times for resetting the audio, which seems to remove the audio disturbances for sometime, but then they started coming again.

We were using Panasonic ergofit headphones for the Fremont users and a professional quality mic (Apogee) for the Austin person. After testing with Jamulus for more than 10-15 hours across several occasions, we have noticed that these two headphone models generally don't produce sound disturbances with people who are in the same metro area:

Panasonic ErgoFit Earbud Headphones with Microphone and Call Controller Compatible with iPhone, Android and Blackberry - RP-TCM125-K - In-Ear (Black), Earpads S/M/L
https://smile.amazon.com/Headphones-Microphone-Controller-Compatible-Blackberry/dp/B00E4LGVUO/ref=sr_1_3?dchild=1&keywords=panasonic+headphones+with+mic&qid=1592108421&sr=8-3

JUKSTG Noise Isolating in-Ear Headphones with Pure Sound and Powerful Bass, Earbuds with High Sensitivity Microphone (for use with Macbook)
https://www.amazon.com/gp/product/B07GK35P2C/ref=ppx_yo_dt_b_asin_title_o00_s00?ie=UTF8&psc=1

The main intent of opening this issue is to resolve ongoing audio disturbances that crop up during a session. It makes having an enjoyable experience very difficult. at the same time, we find that Jamulus technology is quite promising and we would like to be get better for the good of the whole world.

**Comment by gilgongo** [comment]
With so many variables involved in hardware, software and networks between players on any given Jamulus session, it's extremely hard if not impossible to track down the cause of many problems without systematic investigation. So it's good that you've attempted it! I'm surprised that earbuds would be a factor here though, but I'm not qualified to know.

I assume all players had covered the basics that we have advised on [the wiki](https://github.com/corrados/jamulus/wiki/Getting-Started), since things like using onboard sound cards, other things running on clients or servers, or people in the house having Zoom calls, etc. will of course affects things. Perhaps we should produce a diagnostic playbook?

**Comment by WolfganP** [comment]
I suggest to check your audio hardware settings (buffers, drivers) first and make sure they work OK locally (ie running a client and a test server on the same machine) and once everything go smooth then shift to the remote server with the same client setup.

Where exactly is your server running, at your home or a data center close to you? 60ms seems a lot of delay for users in the same area (I have a private server running at home on a raspberry, and my overall delay with a shitty laptop audio on asio4all is 18ms; a friend on the same ISP goes up to approx 40ms).

The folks at World Jam Online (https://www.youtube.com/watch?v=xmU8i0DGLEY) use an UK server for the european folks (diff countries, diff ISPs) and lately a band from the US also played online with a US based server (I believe they are hosted at AWS), and their results are very good.

Check also the forum for tips and other users experiences https://sourceforge.net/p/llcon/discussion/

**Comment by gvermag22** [comment]
Thanks for the thoughts. Not sure where you are located and what the
relative ISP speeds there. I’m in the San Francisco Bay Area and am using a
Zytel solution for powerline Ethernet for the LAN connection.

When I use a private server on jamulusOs (linux) at my home and use the
internal subnet IP (e.g. 10.0.1.100), the overall delay is 20ms. If I use
the external IP for my server, the overall delay is 65-70ms. I have no
control over the delays. It's decided by the ISP.

However, a person I'm working with setup a server on a hosted machine in a
San Franscisco based data center and on connecting to that, we were getting
40ms of overall delay. By the way, we just use the Auto buffer settings and
it seems to work good for the most part. When the musicians are in the same
metro area, using the Panasonic ErgoFit has shown to work mostly. I have
also seen that on some evenings, the audio experience is different with the
SAME setup across users, and this seems to be due to the state of the
internet. The overall delays are similar though.

I have documented all our experiences and setup in these guides:

https://bit.ly/onlinegroupkirtansolutionoverview
https://bit.ly/onlinegroupkirtanforadmins
https://bit.ly/onlinegroupkirtanforusers

Our musical instruments are a bit different from the regular western
instruments. For percussion, we use this instrument
https://www.youtube.com/watch?v=Wx1VdhM1W8k and for singing along we use
this https://www.youtube.com/watch?v=EJ8FzBj2Jbg. These are the two primary
instruments. They are not electric.

Now, we were trying to make a music session work with another member in
Austin, TX. He has a high end professional mic, which is great. We were
using the cheap $15 panasonic ErgoFit headphones and mic. We experienced
audio disturbances now (sputtering, rain, sparking sounds) and then and had
to reconnect to Jamulus many times over. Another issue I've seen is that if
the network configuration (e.g. from Wifi to LAN) of any person changes,
then EVERYONE has to reconnect back to avoid audio disturbances. This is
obviously a bug and needs to be addressed.

**Comment by gilgongo** [comment]
Thanks for providing the links to the documents. Some observations from those:

1. Looking at the [diagram in this document](https://docs.google.com/document/d/1APPYPauKpwQn_SYrbdBCQJyuJ4OX1vdDODgWuh05408/edit), I would think there's quite a high chance of you running out of bandwidth if you have the Jamulus server on a home broadband connection while also having a Zoom audience and multicasting to Facebook or YouTube. I don't know if you've seen [this page](https://github.com/corrados/jamulus/wiki/Quality,-delay-and-network-bandwidth), but I think you may need to check how much upload bandwidth you have on the server's connection. Note from that page that the clients' audio quality setting will determine whether you get problems with bandwidth at the server end too. 

2. It's highly recommended that clients [use external audio interfaces](https://github.com/corrados/jamulus/wiki/Getting-Started#highly-recommended-an-audio-interface). Many on-board sound cards have too much latency or cannot be set to good buffer sizes to avoid audio problems. The Windows ASIO4all driver is OK, but not great in most cases when using onboard sound cards in my experience. Have a look at the [hardware page](https://github.com/corrados/jamulus/wiki/Sound-Devices).

3. Similarly for CPU usage. I would not want to run Zoom and Jamulus on the same machine as the former will want a lot of CPU and you'll see buffer overruns and audio problems because of it. Many Windows machines in particular are also loaded up with anti-virus/malware systems and other incidental software that runs in the background quite causing CPU spikes. If your machine starts doing an anti-malware check in the middle of a session, you will probably get audio problems. 

In short, on the evidence of the information you have provided in those documents, I think your observed problems may be due to overloading the systems involved. Successful Jamulus sessions almost always have nothing but Jamulus running on the server and the clients, and enough bandwidth at the server end such that the combination of all the clients do not max it out.

**Comment by gvermag22** [comment]
Thanks for the detailed note pointing out the sources of troubles.

1. At home, with powerline ethernet, right now at 2:30pm PST on a friday, my ping is 17ms, upload is 95mpbs and upload is 5.6mbps. The requirements for Jamulus is 1mbps up or down tops. The bandwidth should not be an issue. 

2. Except for one person in the session, no one was using an audio interface. He used the Apogee duet. It also seems that the audio interface is only required to minimize delay between the instrument and your own hearing, but we are not using any electronic instrument, so it doesn't seem applicable?

3. The point on the CPU priority/context switching was also in our mind recently, and seems very relevant. It seems that in most situations, we may need to run Zoom and Jamulus on the same machine (it's too much to ask of people to use separate machines for zoom and jamulus, and we're trying to increase adoption of this solution), so we can get around that by increasing the priority of Jamulus process to the maximum (after system background processes). This seems possible on Mac and Linux. 

Windows is a terrible mess due to ASIO drivers, firewalls and anti-virus, so I don't really recommend using Windows. Our experience with mics was also not good with windows. The same mics worked on macbook or linux or with a phone just fine.

By the way, we had also experienced some audio gaps in the zoom to jamulus bridging in a few sessions, and suspect that it was due to CPU prioritization/process context switching.

4. I agree that using the powerline ethernet can introduce packet loss issues. We did face it recently, and had to shift from the corner room of the house to a middle one, which resolved the issue. It just happened all of a sudden. Depending on where you quiet room is in the house, you may very well end up using a powerline ethernet. From our testing, we have seen a drastic difference in delay stability by using powerline ethernet over wifi. With Wifi, the sound used to buffer a lot, and the delay variation was a lot more.

We are going to try with closing all other applications and increasing the Jamulus priority to max on Macbooks and will report our progress.

**Comment by gilgongo** [comment]
In point 1, are you referring to the network that the server is on? In which case 1Mb/s will almost certainly not be enough. How many musicians are you typically hosting on the server BTW?

Regarding point 2, you will need to plug a mic or a pickup in to something in order for Jamulus to transmit the sound of the acoustic instrument to the server. That's what I mean by "audio interface". If you plug the mic into the on-board mic input of most computers, you usually get more latency as computers aren't designed to process audio very well. This isn't an iron rule of course - Mac might be better - but as a general guideline it's worth noting.

As to point 3, you can of course try to alter the priority, but ultimately it's a much better idea not to have the computer that's running Jamulus also running a significantly hungry application as well. 

As to point 4, that's as maybe, but it's another potential source of problems.

In summary: from what you describe of your current setup, I would expect quite a few problems to occur. The [four main ingredients](https://github.com/corrados/jamulus/wiki/Getting-Started) for a good Jamulus session are pretty simple really: wired Ethernet, a good audio interface, no interruptions, and try not to listen to your own sound. Get those right and you will enjoy the standard of sound that the rest of us get. Which is pretty much fine 95% of the time.

**Comment by gvermag22** [comment]
The home server I'm running is on I7 or I9 Intel processor, 8-16GB RAM, SSD hard disk, has 95mbps down, 6mbps up, and runs Zoom and Jamulus server on JamulusOS (linux). It shouldn't be bandwidth or CPU challenged, but can definitely benefit from the higher prioritization of the Jamulus process. 

The number of participants is usually 3-4. Here is a sample session that went very well, https://www.youtube.com/watch?v=7iFw-jmoIVM&feature=youtu.be but the same setup produces different results at different times. The mics we are using are these https://smile.amazon.com/Headphones-Microphone-Controller-Compatible-Blackberry/dp/B00E4LGVUO/ref=sr_1_3?dchild=1&keywords=panasonic+headphones+with+mic&qid=1592108421&sr=8-3 or https://www.amazon.com/gp/product/B07GK35P2C/ref=ppx_yo_dt_b_asin_title_o00_s00?ie=UTF8&psc=1. 

The ambient sound of the musical instruments is getting picked up these headphone mics and that's good enough for our sessions. We have had good recordings with this approach and didn't need to explicitly connect a mic for the instrument. One of our members has bought https://smile.amazon.com/gp/product/B01J7P0OGI/ref=ppx_yo_dt_b_asin_title_o00_s00?ie=UTF8&psc=1 for audio interface and is going to try it out.

**Comment by gilgongo** [comment]
OK, so server bandwidth and CPU are unlikely to be an issue then. Also, just looking at that YouTube video you posted, when you say "participants" do you mean audience over Zoom as well as the performers? The video implies a single (LAN) client for the musicians and Zoom for the audience. That's obviously different from a normal Jamulus session where multiple remote (WAN) musicians play together. But I think you mean both scenarios, looking at your initial post. 

One thing we've not talked about is the apparently quite long delay times you quote in your original post: "fremont folks was 60ms and the delay from Austin was 80-85ms". Are these the ones reported by Jamulus or are you measuring them in another way? Anything much more than 50ms tends not to make a session viable. The only thing you can do about that is to try to find a public server that has a lower ping time for all of you and play on that server instead. This is one of the advantages of public servers over private ones. If you don't like the one you're in, you can move :-) 

It would be interesting to see if that audio interface you mention helps. It looks like a jack to USB adaptor (for PCs without jack inputs) rather than an actual DAC, so may in fact introduce more latency, not less. BTW one way to get a rough idea of that is to run a server on the same machine as the client and then connect to localhost and note what the Jamulus client says the delay is at differ buffer sizes. That delay should be representing your "ambient" delay in signal processed by the machine, give or take a ms or so.

**Comment by gvermag22** [comment]
In the youtube video I posted, while only one Jamulus singer (the lady and
her husband) was singing and playing instruments. The other two (me and a
friend) were listening in on Jamulus to see the audio quality. All 3 of us
were also on Zoom for the video part only. The audio from Jamulus was
getting redirected to Zoom via QJackCtl utility on the server (the audio
bridging steps are mentioned in https://bit.ly/onlinegroupkirtanforadmins
clearly. Other than this test case, usually, our setup is that we have 3-4
singers/musicians on Jamulus and 10-20 zoom attendees to listen into the
performance.

I would very much like to figure out the longer delay times. We think the
solution definitely works, but is flaky at times, and that's a concern for
us. Sometimes, too many variables change here and there. Anyhow, without
the cheap audio interface/adapter that I mentioned, on a macbook, the
overall delay for localhost (127.0.0.1) Jamulus connection (server is also
on the same machine) is 20-22ms. Buffer delay is 5.33ms and the Jitter
Buffer is Auto. We are always using Auto for the Jitter Buffer to avoid too
much confusion/setup overhead. On a Linux machine, where the audio bridging
is enabled via QJackCtl, the overall delay is 36ms. The LED is green (but
right now no Jamulus users are connected to it except the local Jamulus
client. The Buffer delay is 10.67ms (this is the only enabled option).
Jitter Buffer is set to Auto.

I don't understand the use case of the Public server. Aren't these like
public bathrooms where anyone can just walk in? We need a gated space so
that our session wouldn't be disturbed. The concept/usage or public server
is confusing to me. We have not tried public servers yet. If we can find a
public server that has delays of <=50ms and that no one gatecrashes (is
there a way to put a do not enter sign) then we would love to use that
option. I feel there should be more documentation and clarification
provided on this.

**Comment by gilgongo** [comment]
The delay on the Macbook seems about right for its built-in sound card. I get about 20ms on the Ubuntu 20.04 Linux box I use for my Jamulus client (an old Core Duo bought from eBay), to which I have a Tascam DR-40 connected in audio interface mode. On your Linux machine, I would guess the increased delay is due to it running with a frame rate of 256. You should be able to set this to 128 in the QJackCtl setup and see if you can reliably get a lower delay on that setting. This will depend on the speed of the CPU though, which in turn may be affected by Zoom (and possibly also the bridging, but I'm not sure).

As to public servers - if your remote players can find one with a better network ping time than your private server, then you can simply join that and solo each other. I agree we don't make a big enough point about that, so I've made this [a bit more obvious](https://github.com/corrados/jamulus/wiki/Running-a-Server) in the docs now. BTW non-soloed people can still text chat to you, but you can of course ignore that if you want. But you never know, they might be able to contribute something :-)

**Comment by gvermag22** [comment]
For some reason, even though the setting in QJackCtl is 128ms for the Intel:PCH interface, only the 256ms setting is enabled in Jamulus settings. I have noticed that the Zoom app for Ubuntu 64bit kind of hangs when it connects the audio initially. The machine is powerful though. But after 20 seconds or so, it becomes responsive.

We did some testing about the ambient delay with 3 test cases on Linux and got the same ambient delays, so that doesn't seem to be a factor:
1) USB Sennheiser PC8 mic: 20ms
2) Panasonic ErgoFit headphones/mic connected to https://smile.amazon.com/gp/product/B01J7P0OGI/ref=ppx_yo_dt_b_asin_title_o00_s00?ie=UTF8&psc=1: 20ms
3) Cardoid XLR mic connected to https://www.amazon.com/Focusrite-Scarlett-Audio-Interface-Tools/dp/B07QR73T66: 20 ms

Thanks for updating the documentation for public servers. I get the approach for checking "solo" for the specific people in our group in the same public server, and get the session going, however, we have concerns about security and privacy. If you can recall all the zoom bombings etc recently in the news, you may get the drift. The "solo your group members" approach seems like huddling in a public bathroom. Others can still hear you and have to specifically solo their group members to do their session. This is not a clean approach. It would be better to let group members define their "rooms" and then lock or unlock their room with a password. Please consider this approach for enhancement.

**Comment by gvermag22** [comment]
Today, we did another session which was recorded on Zoom. Here was the setup:

1. My server is running on JamulusOS on a bootable USB stick on a windows machine. This server has Jamulus Server and Zoom bridging. Delay times: 35ms. Jitter Buffer: Auto. Buffer delay: 256ms. I bumped up the cpu priority of Jamulus server and client (for bridging) to max using renice command ($ sudo renice -n -20 -p <pid>). Connected to LAN via powerline ethernet.
2. My Jamulus client was running on Macbook pro CatalinaOS. Delay 50ms. Jitter buffer:Auto. Buffer delay:128ms. Connected to LAN via powerline ethernet. Mic: Panasonic ErgoFit ($15 on amazon).
3. Another musician with similar setup but direct LAN connection, and overall delay of 40-60ms.

Here is the recording https://youtu.be/1o8YyNHgU9o. There were audio disturbances at 5:08, 6:17, 6:55, 7:00-7:09, 1:04:40, 1:13:58, 1:23:55. AFter the 10 min mark or so, I reconnected via internal LAN subnet 10.0.1.100, instead of the external IP, and that reduced my overall delay to 20ms. So, there are different kind of audio issues: buffering (when the sound becomes too rich), sparking sounds and sputtering.

While the overall program was still ok, but the audio quality was not perfect. I suspect it is due to the cheap mics we are using. I tried at least 5-6 different mics from Amazon and returned them after extensive testing with other people here. I found audio issues with: https://smile.amazon.com/gp/product/B07Z4RHL2G/ref=ppx_yo_dt_b_asin_title_o04_s00?ie=UTF8&psc=1, https://smile.amazon.com/gp/product/B075KL6ZLC/ref=ppx_yo_dt_b_asin_title_o09_s00?ie=UTF8&psc=1, https://smile.amazon.com/gp/product/B005HWEZGG/ref=ppx_yo_dt_b_search_asin_title?ie=UTF8&psc=1 (beep beep life support system kind of sound and sparking kind of sound periodically). Selecting a decent mic that doesn't break the bank has been the most frustrating experience of using Jamulus, but then the solution has so much promise that I have persevered in the hope of finding the right solution.

Our temple priest has used this audio interface with two mics and the sound is very clear, but it seems overkill for our needs: https://www.bhphotovideo.com/c/product/1479273-REG/focusrite_scarlett_solo_3g_scarlett_solo_usb_audio.html

We only need one mic per musician, which can be an omnidirectional one. We don't use electronic instruments, so an audio interface seems unnecessary as there is no instrument delay to reduce. That sound is blended into the singing sound automatically due to proximity, and it's good enough. Which semi-professional mic can you recommend off amazon that can work well with Jamulus. I'm skeptical of buying another off other sites due to sketchy return policies/experience.

**Comment by gilgongo** [comment]
> For some reason, even though the setting in QJackCtl is 128ms for the Intel:PCH interface, only the 256ms setting is enabled in Jamulus settings. 

I'm not familiar with the issues around bridging to Zoom using Jack I'm afraid so I don't know if that's expected behaviour. Does it do the same if you don't bridge?

> We did some testing about the ambient delay with 3 test cases on Linux and got the same ambient delays

Yes, I would expect all types of input to the same sound card would have the same delay, since the delay is almost all from the sound card's hardware and driver capabilities. 

> It would be better to let group members define their "rooms" and then lock or unlock their room with a password. Please consider this approach for enhancement.

The idea of password protected private/public servers has been discussed in the past here #299. I understand your concerns about using a public server - I merely point out their use as a means of solving one of your problems around sound quality.  

> Today, we did another session which was recorded on Zoom.

In 1. I'd expect some problems with running a Jamulus server, a client and Zoom running on the same machine, although I assume the client isn't sending any sound to the server, a buffer delay of 256ms is (as the manual says) not generally a good idea if you want to avoid glitches unless on a very slow computer, or with a slow internet connection (neither of which you have?) So that's a bit of a mystery at least as to why you can't run the client in a more stable mode. 

> I reconnected via internal LAN subnet 10.0.1.100, instead of the external IP, and that reduced my overall delay to 20ms.

Yes, as we state in the docs. Don't use a WAN address if you can use a LAN one.
 
> an audio interface seems unnecessary as there is no instrument delay to reduce. 

That's true in the case of USB microphones (I've recently clarified the docs on that point) because they will do the digital/analogue conversion before it reaches Jamulus. But it's not true of analogue mics. However, as you have all your musicians in one physical space the delay in this case may not matter?

Blue Yeti mics have been mentioned as working pretty well (see the [known good hardware](https://github.com/corrados/jamulus/wiki/Sound-Devices)list). 

Lastly, at the risk of asking a silly question here - given the fact that you have all your musicians in the same physical location, why are you using Jamulus at all? Is the Zoom audio worse? Apologies, as I'm not experienced with it enough to know why you'd use Zoom for video and Jamulus for audio, given the inherent complexity in doing that.

**Comment by gvermag22** [comment]
It turns out that having 10.57ms delay (256) on Jamulus server is advantageous to the overall delay (35ms). With 5.33ms (128), it rises to 55ms. I had to restart QJackCtl and Jamulus to enable the buffer delay option in Jamulus settings.

It's becoming clear that the overall delay is lesser on weekends as compared to weekdays. This seems natural as there is more work related traffic on weekday evenings. Thus, the same setup can produce a degraded experience on weekdays.

It's also becoming clear that by running Jamulus server on a hosted machine reduces overall delay by 15ms or so due to lesser hops to the internet backbone. We tried this a couple times. However, this has the additional overhead of renting the virtual hosted server, and is not for everyone. If I use LAN IP 10.0.1.100, my overall delay is 35ms, but others may face longer delays due to external IP access. If it's a weekend, that challenge is reduced a bit due to less internet traffic.

The recording I posted was a simple use case where all musicians are in one spot. But I was also singing for the response and was in a different location. We found that the Jamulus audio quality is much much better than Zoom's audio (when all musicians are in the same location), even after tweaking zoom's settings for music https://docs.google.com/document/d/1VSO5mbWtzVD-_-XN_pwPzL4eUUv1E3DSxNqD-3a-3T0/edit?usp=sharing . Our complex setup is when instrument players/musicians are in different locations, and you can see the recording for it at https://www.youtube.com/watch?v=si77KW1oLJk&list=PLW6aF2mZXK_nc4EjOO79DCnjvYiQ_VN4o&index=2&t=2s. 

In the above recording, you'll see that the drum player was having a time sync challenge initially. I've realized that we need to hear the Jamulus output instead of hearing ourselves via the earplugs in our ears. When something is in your ear and you say something, you hear it first. We need USB headsets (and USB mics) that will more prominence to the Jamulus output stream for hearing. I know you mentioned that Blue Yeti has been reported to work with Jamulus, but it's pretty expensive for amatuer singers like us. We will try a few models and let you know which ones work.

On the Jamulus server, the jamulus client is only used for redirecting the stream to zoom via QjackCtl. The actual Jamulus client runs on a separate machine.

**Comment by sthenos** [comment]
I feel like jumping in here and stating some obvious things.  I run the current World Jam weekly events that are streamed online to facebook and youtube and whilst we do at times have problems with individual musicians and singers jittering on their line, the overally sound going out to facebook/youtube is of high quality.

1.  Firstly servers are all runnning on cloud VM's with at least 8MB ram on Unbuntu and are dedicated either 2 or 4 vCPU's, but number of cpu's is not important for jamulus as it is not multithreaded.

2.  Secondly our audio stream is fed from this server into a dedicated cloud VM (now at least) windows server desktop with 8 cpu's and 32GB of ram.  This, we have found and with helpful advice from another fellow streamer, is the minimum for streaming audio and video out anywhere.     The audio for this jamulus sesssion is set at 1024 buffers (i.e. much higher than the buffer a typical player needs).  Latency is not a problem for streaming, quality is more important.    The audio is delayed by about 250ms deliberately to match up with the ~300ms delay that zoom has built in.   This is then stream out to multiple sources.

3. For all our musicians, the absolutely requirement is 1.  an ethernet cable, powerline or wifi is not sufficient.  2.  An external audio interface, internal soundcards are useless particularly for drummers and anyone needing to tight sound.   3. A fast PC in order to have the buffer set on 64 samples.     These are the minimum requirements for musicians to get the best out of jamulus.  

We find with these settings that dropouts are minimised, sound quality is high, so is the zoom video quality.

**Comment by gilgongo** [comment]
@sthenos Thanks for this, so for @gvermag22 would I be correct in thinking the main bottleneck for him to look at in his setup is the streaming to Zoom and Facebook, if that needs a minimum of an 8 CPUs machine with 32GB of RAM? Interesting that the audio buffer for the client on that needs to be 1024 too.

**Comment by sthenos** [comment]
I think the difference is I'm streaming in 1080p to restream using OBS at a very high bitrate for good quality.   I'm also running zoom too.  I don't believe from the thread above, they are streaming to zoom per se, just having it open.   I guess they would want a much lower buffer sample amount so that the zoom audio is in sync with the video.   We don't have that requirement as we just delay our audio to match zooms bad latency.   Having 1024 means that we don't have any drop outs on the live audio stream, but would not be appropriate for the fella above with his setup.  Streaming the audio to zoom is going to have a problem with out of sync audio and video.     Having zoom open with low buffer samples WILL absolutely impact audio quality without a doubt. Even on the 8 cpu, 32GB machine having zoom and OBS open on a buffer sample size of 512 caused dropouts!

**Comment by gvermag22** [comment]
@gilgongo @sthenos thank you for the excellent insights. 

From your comments, my takeaways (and comments) are:

1) Interesting to know Jamulus is not multi-threaded, so # of CPUs don't make a difference (yet). However, CPU priority should make a difference. We tried it ($sudo renice -n -20 -p <pid>) and at least we didn't lose anything. It would have only helped. We were experiencing some audio gaps in between, and the last time we did the session, I didn't hear audio gaps.

2) Having a high buffer delay setting leads to better quality audio. We had also noticed that by increasing to 256 from 128, the audio quality had become smoother. I didn't realize you could set it all the way up to 1024. We are using JamulusOS for our server, which is linux. I just setup 1024 frames/period in QJackCtl and saw that the Jamulus connection to the server has 100ms delay. Of course, I didn't test with other musicians with this setup yet, but it does seem counter intuitive that a higher delay would make audio better. Will keep you posted on what we find.

3) you're recommending external audio interfaces as required for musical instruments, but I'm guessing the instruments are electronic in this use case. Our instruments are not electronic and their ambient sound has been good enough in our recordings. Also, audio interface is useful for XLR compatibility mics and for converting analog to digital. However, USB based mics are already digital, and cheaper. We are not rich musicians with a lot of money to spare. We might test with some USB based gaming headsets with cardioid pickup pattern and will report progress.

4) there will be a visual lag between zoom video and audio (when jamulus is streamed to it), but that is expected, and we were good with it. people can hardly make out the difference. It's fascinating that you are able to almost sync up zoom audio and video by delaying the audio stream. That's a cool approach.

As a reminder, our setup is outlined in https://bit.ly/onlinegroupkirtansolutionoverview. Our main issue is with audio disturbances (buffering, sputtering, sparking, garbling as if it came from space). But then, right now, we are using the cheapest (and tested) mics ($15) possible that are working 85-90% of the time with some audio glitches, but we want to make the presentation flawless. Most importantly, we want to sync up the drum/percussion instrument with the singer in timing. We realize that the Jamulus software is flaky currently and needs audio improvements.

**Comment by sthenos** [comment]
hi @gvermag22 

Reponses to some of your items:

> Interesting to know Jamulus is not multi-threaded, so # of CPUs don't make a difference (yet). However, CPU priority should make a difference. We tried it ($sudo renice -n -20 -p ) and at least we didn't lose anything. It would have only helped. We were experiencing some audio gaps in between, and the last time we did the session, I didn't hear audio gaps.

I'd be very suprised if CPU priority would make any difference to jamulus if you are also running Zoom.  Zoom is incredibly processor hungry and for that reason we run it on a different machine from the instance where Jamulus is running for a musician/singer to play in low latency.

> I didn't realize you could set it all the way up to 1024. 

You can set it to whatever you want, Jamulus will just use the buffer sample size that the underlying device is using.

> You're recommending external audio interfaces as required for musical instruments, but I'm guessing the instruments are electronic in this use case. 

Absolutely not, I have a Presonus Studio 1824c and combination of electronic instruments (electric guitars/keyboards) as well as microphones for my voice and pickups on acoustic guitars, etc.   USB mics are inherently poorly performing, they often don't have dedicated drivers for them.   

As a reminder, our setup is outlined in https://bit.ly/onlinegroupkirtansolutionoverview. Our main issue is with audio disturbances (buffering, sputtering, sparking, garbling as if it came from space). But then, right now, we are using the cheapest (and tested) mics ($15) possible that are working 85-90% of the time with some audio glitches, but we want to make the presentation flawless. Most importantly, we want to sync up the drum/percussion instrument with the singer in timing. We realize that the Jamulus software is flaky currently and needs audio improvements.

Jamulus software is not flaky.  Over the past 3 months I've found Jamulus software on both client and server to be absolutely possible to create professional audio recordings which we are using on tracks being released to a wide audience.  What IS flaky is the computer equipment and audio equipment that some people are using, the network congestion on the route to the server and running servers on desktops that are not designed for that purpose.   Run a jamulus server on an actual cloud VM server designed for that purpose and have high quality audio equipment, fast media PC and good network connectivity and ethernet connection and you'll have a massively improved experience.   For instance I was playing today all through my lunch break on an 8 cpu 32GB ram GCP cloud E2 VM and did not receive a single pop or dropout and perfectly in time with a drummer who was in Italy. (I'm in the UK).

Simon

**Comment by ignotus666** [comment]
@gvermag22

> Having a high buffer delay setting leads to better quality audio. We had also noticed that by increasing to 256 from 128, the audio quality had become smoother. I didn't realize you could set it all the way up to 1024. We are using JamulusOS for our server, which is linux. I just setup 1024 frames/period in QJackCtl and saw that the Jamulus connection to the server has 100ms delay. Of course, I didn't test with other musicians with this setup yet, but it does seem counter intuitive that a higher delay would make audio better. Will keep you posted on what we find.

It sounds better because you're giving the device more time to process the audio and thus reduce the possibility of dropouts. But what you really want is to be able to keep the buffer value low to reduce latency while also having smooth audio, which will be pretty hard to achieve using either onboard audio or USB mics.

> you're recommending external audio interfaces as required for musical instruments, but I'm guessing the instruments are electronic in this use case. Our instruments are not electronic and their ambient sound has been good enough in our recordings. Also, audio interface is useful for XLR compatibility mics and for converting analog to digital. However, USB based mics are already digital, and cheaper. We are not rich musicians with a lot of money to spare. We might test with some USB based gaming headsets with cardioid pickup pattern and will report progress.

You **will** benefit from using an external interface. It doesn't matter if you're singing into an analog mic or using an electronic keyboard. There still has to be A/D and/or D/A conversion either way and an external device will be better/faster at it than the onboard device. Using a USB mic is even worse than using an analog one connected straight to the PC because you force the PC to collect audio from one device (USB mic) and output it through another (onboard audio device), neither of which are at all suitable for low latency audio. What's more, the A/D converter in such USB mics is most probably even worse than what's in your PC. Also, mic/headphone/earphone quality has no impact on latency or audio dropouts whatsoever - more expensive gear might sound better in terms of hiss or frequency response, but will do nothing to improve latency or dropouts (garbled sound). You don't need to buy an expensive interface either - there are many cheap USB interfaces (check the Jamulus wiki) that will work. I think you'd be better off investing in one than in USB gaming headsets which IMHO will be a waste of money.

The 3 requirements that @sthenos outlined above (ethernet cable, external device, reasonably fast PC) are pretty much unavoidable; for normal use with just audio I would say numbers 1 and 2 are the most critical, but if you also want video streaming number 3 will also be a must. From what I gather, you are failing to fulfil at least numbers 1 and 2, so blaming your problems on Jamulus supposedly being "flaky software" is misguided at best.

My 2 pfennigs.

**Comment by gvermag22** [comment]
We have tried with Focusrite Scarlettt external audio interface also and professional cardioid XLR mics also, and the delay has remained as 40-45ms—the same as that with USB headsets, so we are not sure of the earlier comments. 

Last night we tried with https://smile.amazon.com/gp/product/B07ZK6JLJM/ref=ppx_yo_dt_b_asin_title_o01_s00?ie=UTF8&psc=1 usb mic+headset and https://smile.amazon.com/Yamaha-CM500-Headset-Built-Microphone/dp/B000BZ6ZX0/ref=sr_1_1_sspa?dchild=1&keywords=yamaha+headset&qid=1593811655&s=electronics&sr=1-1-spons&psc=1&spLa=ZW5jcnlwdGVkUXVhbGlmaWVyPUFHUlBJQkk4WFVZVEomZW5jcnlwdGVkSWQ9QTA5MDk0ODQxMlMxQjJEU0dLWjNFJmVuY3J5cHRlZEFkSWQ9QTA5OTc3OTAzSVM4SUROUlZERVFIJndpZGdldE5hbWU9c3BfYXRmJmFjdGlvbj1jbGlja1JlZGlyZWN0JmRvTm90TG9nQ2xpY2s9dHJ1ZQ==. The results were very good audio (will be testing more tonight for thoroughness). One person was using macbook & Razer Kraken mic/headset. Other person used LInux + Yamaha mic/headset & a small $10 usb/3.5mm interface.

Another case in point is this recording https://youtu.be/1kI2dzd7Thw?t=2023. The person with white headset used a Blue Yeti USB mic. He was singing and playing a percussion instrument. Another singer used a Sennheiser PC8 mic/headset with Linux. Another musician used regular headphones and played the harmonium. Their delay was ~50ms. None of them used an external audio interface. They were using Zoom for the video as well. While it's not a professional level performance, the recording is ok for the most part, except the occasional jitters and the spiking of audio from the percussion instrument. Even if an external audio interface was used, it's unclear what would be the difference in the audio experience. What are your thoughts?

From our testing/experience, the overall delay is mostly decided by the time of the day (evening vs daytime), the day (weekend vs weekday), and server location (home LAN vs data center). 45-50ms is achievable on weekday evenings if the jamulus server is running on data center. Powerline adapter is unavoidable sometimes due to room location in the house, but it does introduce occasional spikes that result in audio garbling/jittering. The quality of the mic affects different audio effects such as background hiss/static, sparking. These are the two broad kinds of audio issues we have seen and can be explained. 

About the flakiness, we have noticed many times that just reconnecting to Jamulus does away with a lot of audio issues—even with the same mic/headset. Also, if the network configuration of one musician changes (e.g. Wifi to LAN), it spoils the experience for everyone else, and everyone has to reconnect. What is that, if not software flakiness? Jamulus has gained traction only recently due to corona virus situation. It's not being developed at the commercial level, which obviously means that the technology has ways to go, but is promising all the same.

**Comment by corrados** [comment]
> Also, if the network configuration of one musician changes (e.g. Wifi to LAN), it spoils the experience for everyone else

I have created a separate issue for this: https://github.com/corrados/jamulus/issues/426

**Comment by gvermag22** [comment]
Thank you. Can you also create another issue for the latest connected
devices not showing up in the jamulus settings device list unless jamulus
is relaunched. Many people don’t get it why the device isn’t showing up
after a mic or headset was connected. It should automatically add the
device entries.

**Comment by corrados** [comment]
There is already an Issue existing for this: https://github.com/corrados/jamulus/issues/181

**Comment by corrados** [comment]
This Issue is about a month old. Are there any open issues left or can I close this one?

**Comment by gvermag22** [comment]
The issues are still pertinent. I will add some more evidence tomorrow.

**Comment by gilgongo** [comment]
Now over three months old...

**Comment by corrados** [comment]
I think it is time to close this Issue. If anything is open, **gvermag22**, please just re-open or create a new Issue.
